sdk/eventhub/event-hubs/src/eventHubReceiver.ts

```diff
@@ -557,39 +578,85 @@ export class EventHubReceiver extends LinkEntity {
     );
   }
 
+  private async _drainAndAddCredits(
+    onMessage: OnMessage,
+    abortSignal?: AbortSignalLike
+  ): Promise<void> {
+    // add explanation
+    await delay(0);
+    this._usingInternalQueue = true;
+    while (this.internalQueue.length) {
+      if (!this._onMessage) {
+        this._usingInternalQueue = false;
+        return;
+      }
+
+      if (abortSignal && abortSignal.aborted) {
+        this._usingInternalQueue = false;
+        return;
+      }
+
+      if (this._onMessage !== onMessage) {
+        this._usingInternalQueue = false;
+        return;
+      }
+      const eventData = this.internalQueue.splice(0, 1)[0];
+      onMessage(eventData);
+      // allow the event loop to process any blocking code outside
+      // this code path before sending the next event.
+      await delay(0);
+    }
+    this._usingInternalQueue = false;
+
+    if (this._onMessage) {
```

Comment:
What do you think of renaming `_drainAndAddCredits` into `_useInternalQueue` and move the code from this line onwards inside `registerHandlers`? I'd like to avoid helper methods do more than 1 thing if possible

Reply:
I think I can refactor this function into two that also lets me manage initializing the link and managing the credits.

First, I'd update `registerHandlers` to accept a `maximumCredits` parameter and a `isStreaming` parameter.

The `_useInternalQueue` method would still contain everything above this line, and also return how many messages were sent.

Then the new helper function that gets called after `_useInternalQueue` is finished would initialize the link with a prefetchCount of 0 if it hasn't been already, and add credits based that takes into account the number of messages already handled. The rules for handling credits (split out for clarity) would be:
#### Batching
```typescript
const creditsToAdd = Math.max(maximumCredits - (this.existingCredits + processedCreditsCount), 0);
this.prefetchCount = 0; // indicates that credits should not be added as events are received
```
#### Streaming
```typescript
// always add at least 1 credit for streaming mode to ensure new messages are being received
const creditsToAdd = Math.max(maximumCredits - (this.existingCredits + processedCreditsCount), 1);
this.prefetchCount = maximumCredits; // indicates that credits should be added as events are received
```

When `clearHandlers` is called, `this.prefetchCount` will be set to 0. `_onAmqpMessage` will also be updated to check if `this.prefetchCount` is 0 to determine if more credits should be added or not when a message is received.